// apollo-server-koa/dist/ApolloServer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const corsMiddleware = require("@koa/cors");
const bodyParser = require("koa-bodyparser");
const graphql_playground_html_1 = require("@apollographql/graphql-playground-html");
const apollo_server_core_1 = require("apollo-server-core");
const accepts = require("accepts");
const typeis = require("type-is");
const koaApollo_1 = require("./koaApollo");
const apollo_upload_server_1 = require("apollo-upload-server");
var apollo_server_core_2 = require("apollo-server-core");
exports.GraphQLExtension = apollo_server_core_2.GraphQLExtension;
const fileUploadMiddleware = (uploadsConfig, server) => (ctx, next) => __awaiter(this, void 0, void 0, function* () {
    if (typeis(ctx.req, ['multipart/form-data'])) {
        try {
            ctx.request.body = yield apollo_upload_server_1.processRequest(ctx.req, uploadsConfig);
            return next();
        }
        catch (error) {
            if (error.status && error.expose)
                ctx.status = error.status;
            throw apollo_server_core_1.formatApolloErrors([error], {
                formatter: server.requestOptions.formatError,
                debug: server.requestOptions.debug,
            });
        }
    }
    else {
        return next();
    }
});
const middlewareFromPath = (path, middleware) => (ctx, next) => {
    if (ctx.path === path) {
        return middleware(ctx, next);
    }
    else {
        return next();
    }
};
class ApolloServer extends apollo_server_core_1.ApolloServerBase {
    createGraphQLServerOptions(ctx) {
        const _super = name => super[name];
        return __awaiter(this, void 0, void 0, function* () {
            return _super("graphQLServerOptions").call(this, { ctx });
        });
    }
    supportsSubscriptions() {
        return true;
    }
    supportsUploads() {
        return true;
    }
    applyMiddleware({ app, path, cors, bodyParserConfig, disableHealthCheck, onHealthCheck, }) {
        if (!path)
            path = '/graphql';
        if (!disableHealthCheck) {
            app.use(middlewareFromPath('/.well-known/apollo/server-health', (ctx) => {
                ctx.set('Content-Type', 'application/health+json');
                if (onHealthCheck) {
                    return onHealthCheck(ctx)
                        .then(() => {
                        ctx.body = { status: 'pass' };
                    })
                        .catch(() => {
                        ctx.status = 503;
                        ctx.body = { status: 'fail' };
                    });
                }
                else {
                    ctx.body = { status: 'pass' };
                }
            }));
        }
        let uploadsMiddleware;
        if (this.uploadsConfig) {
            uploadsMiddleware = fileUploadMiddleware(this.uploadsConfig, this);
        }
        this.graphqlPath = path;
        if (cors === true) {
            app.use(middlewareFromPath(path, corsMiddleware()));
        }
        else if (cors !== false) {
            app.use(middlewareFromPath(path, corsMiddleware(cors)));
        }
        if (bodyParserConfig === true) {
            app.use(middlewareFromPath(path, bodyParser()));
        }
        else if (bodyParserConfig !== false) {
            app.use(middlewareFromPath(path, bodyParser(bodyParserConfig)));
        }
        if (uploadsMiddleware) {
            app.use(middlewareFromPath(path, uploadsMiddleware));
        }
        app.use(middlewareFromPath(path, (ctx, next) => {
            if (this.playgroundOptions && ctx.request.method === 'GET') {
                const accept = accepts(ctx.req);
                const types = accept.types();
                const prefersHTML = types.find((x) => x === 'text/html' || x === 'application/json') === 'text/html';
                if (prefersHTML) {
                    const playgroundRenderPageOptions = Object.assign({ endpoint: path, subscriptionEndpoint: this.subscriptionsPath }, this.playgroundOptions);
                    ctx.set('Content-Type', 'text/html');
                    const playground = graphql_playground_html_1.renderPlaygroundPage(playgroundRenderPageOptions);
                    ctx.body = playground;
                    return next();
                }
            }
            return koaApollo_1.graphqlKoa(this.createGraphQLServerOptions.bind(this))(ctx, next);
        }));
    }
}
exports.ApolloServer = ApolloServer;
exports.registerServer = () => {
    throw new Error('Please use server.applyMiddleware instead of registerServer. This warning will be removed in the next release');
};
